### PR TITLE
Implement WebCodecsVideoDecoder with VPx backend

### PR DESCRIPTION
**Pull request by youennf**
#### db7df81e7dc760e7dbabec0093f07d1549330dd0
<pre>
Implement WebCodecsVideoDecoder with VPx backend
<a href="https://bugs.webkit.org/show_bug.cgi?id=245878">https://bugs.webkit.org/show_bug.cgi?id=245878</a>
rdar://problem/100609725

Introduce VideoDecoder interface that is used by WebCodecsVideoDecoder as the interface to actual decoders.
Implement WebCodecs VideoDecoder algorithms.
Add VP8 and VP9 decoder implementations for VideoDecoder.

Reviewed by Eric Carlson.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/video-decoder.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/video-decoder.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.worker_vp8-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_vp8-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.window-expected.txt:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOS.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOSsim.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.mac.exp:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.h:
(WebCore::WebCodecsEncodedVideoChunk::byteLength const):
(WebCore::WebCodecsEncodedVideoChunk::data const):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::WebCodecsVideoDecoder::configure):
(WebCore::WebCodecsVideoDecoder::decode):
(WebCore::WebCodecsVideoDecoder::flush):
(WebCore::WebCodecsVideoDecoder::reset):
(WebCore::WebCodecsVideoDecoder::close):
(WebCore::isValidDecoderConfig):
(WebCore::WebCodecsVideoDecoder::isConfigSupported):
(WebCore::WebCodecsVideoDecoder::closeDecoder):
(WebCore::WebCodecsVideoDecoder::resetDecoder):
(WebCore::WebCodecsVideoDecoder::scheduleDequeueEvent):
(WebCore::WebCodecsVideoDecoder::setInternalDecoder):
(WebCore::WebCodecsVideoDecoder::queueControlMessageAndProcess):
(WebCore::WebCodecsVideoDecoder::processControlMessageQueue):
(WebCore::WebCore::WebCodecsVideoDecoder::suspend):
(WebCore::WebCodecsVideoDecoder::stop):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoderConfig.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoderConfig.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoderSupport.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/LibWebRTCVPXVideoDecoder.cpp: Added.
(WebCore::vpxQueue):
(WebCore::LibWebRTCVPXInternalVideoDecoder::create):
(WebCore::LibWebRTCVPXInternalVideoDecoder::postTask):
(WebCore::LibWebRTCVPXInternalVideoDecoder::close):
(WebCore::LibWebRTCVPXVideoDecoder::LibWebRTCVPXVideoDecoder):
(WebCore::LibWebRTCVPXVideoDecoder::~LibWebRTCVPXVideoDecoder):
(WebCore::LibWebRTCVPXVideoDecoder::decode):
(WebCore::LibWebRTCVPXVideoDecoder::flush):
(WebCore::LibWebRTCVPXVideoDecoder::reset):
(WebCore::LibWebRTCVPXVideoDecoder::close):
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
(WebCore::createInternalDecoder):
(WebCore::LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder):
(WebCore::LibWebRTCVPXInternalVideoDecoder::Decoded):
* Source/WebCore/platform/LibWebRTCVPXVideoDecoder.h: Added.
* Source/WebCore/platform/VideoDecoder.cpp: Added.
(WebCore::VideoDecoder::create):
* Source/WebCore/platform/VideoDecoder.h: Added.
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProviderCocoa.cpp:

Canonical link: <a href="https://commits.webkit.org/255138@main">https://commits.webkit.org/255138@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/69d9e3404cc99838567d559ba6b446c6f46cdb8a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91449 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22099 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101175 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161181 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/557 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83797 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97538 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97107 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/378 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27338 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82302 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70371 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35583 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15988 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33369 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17079 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3572 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37169 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39089 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36196 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    m_state = WebCodecsCodecState::Configured;
+    m_isKeyChunkRequired = true;
+    
+    queueControlMessageAndProcess([this, config = WTFMove(config), identifier = scriptExecutionContext()->identifier()]() mutable {
+        m_isMessageQueueBlocked = true;
+        VideoDecoder::PostTaskCallback postTaskCallback;
+        if (isMainThread()) {
+            postTaskCallback = [](auto&& task) {
+                callOnMainThread(WTFMove(task));
+            };
```
Comment (eric-carlson) [inline Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp]: Is the correct, you want to `callOnMainThread` if you are _already_ on the main thread?

**Review by eric-carlson — APPROVED** [review]
r=me

File: Source/WebCore/platform/VideoDecoder.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    struct DecodedFrame {
+        Ref<VideoFrame> frame;
+        int64_t timestamp { 0 };
+        std::optional<uint64_t> duration;
+    };
+    using CreateResult = Expected<UniqueRef<VideoDecoder>, String>;
+
+    using PostTaskCallback = Function<void(Function<void()>&&)>;
+    using OutputCallback = Function<void(DecodedFrame&&)>;
+    using CreateCallaback = CompletionHandler<void(CreateResult&&)>;
```
Comment (philn) [inline Source/WebCore/platform/VideoDecoder.h]: Spurious `a` in `CreateCallaback`

File: Source/WebCore/platform/VideoDecoder.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+class VideoFrame;
+
+class VideoDecoder {
+public:
+    virtual ~VideoDecoder() = default;
+
+    struct EncodedFrame {
+        Span<const uint8_t> data;
+        bool isKeyFrame { false };
+        int64_t timestamp { 0 };
```
Comment (philn) [inline Source/WebCore/platform/VideoDecoder.h:46]: pardon my naive question, what is the semantics for this? presentation ts or decode ts?

File: Source/WebCore/platform/VideoDecoder.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        std::optional<uint64_t> duration;
+    };
+    struct DecodedFrame {
+        Ref<VideoFrame> frame;
+        int64_t timestamp { 0 };
```
Comment (philn) [inline Source/WebCore/platform/VideoDecoder.h:51]: This is decode ts, right?

File: Source/WebCore/platform/VideoDecoder.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        std::optional<uint64_t> duration;
+    };
+    using CreateResult = Expected<UniqueRef<VideoDecoder>, String>;
+
+    using PostTaskCallback = Function<void(Function<void()>&&)>;
+    using OutputCallback = Function<void(DecodedFrame&&)>;
+    using CreateCallaback = CompletionHandler<void(CreateResult&&)>;
+    static void create(const String&, CreateCallaback&&, OutputCallback&&, PostTaskCallback&&);
+
+    using DecodeCallaback = Function<void(String&&)>;
```
Comment (philn) [inline Source/WebCore/platform/VideoDecoder.h]: Ditto

File: Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp (hunk 1 above)
Comment (youennf) [reply Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp]: Right, let's change to ensureOnMainThread.

File: Source/WebCore/platform/VideoDecoder.h (hunk 2 above)
Comment (youennf) [reply Source/WebCore/platform/VideoDecoder.h]: Will fix

File: Source/WebCore/platform/VideoDecoder.h (hunk 3 above)
Comment (youennf) [reply Source/WebCore/platform/VideoDecoder.h:46]: It is presentation ts: https://w3c.github.io/webcodecs/#dom-encodedvideochunk-timestamp-slot

File: Source/WebCore/platform/VideoDecoder.h (hunk 5 above)
Comment (youennf) [reply Source/WebCore/platform/VideoDecoder.h]: Will fix

File: Source/WebCore/platform/VideoDecoder.h (hunk 4 above)
Comment (youennf) [reply Source/WebCore/platform/VideoDecoder.h:51]: It is the timestamp given as input by the web page, the presentation timestamp.

Comment (youennf) [reply Source/WebCore/platform/VideoDecoder.h:51]: Maybe we should name it as presentationTimestamp.

**Pull request merged**